rust/lance/src/dataset/write/delete.rs

```diff
@@ -66,88 +70,195 @@ async fn apply_deletions(
+/// Builder for configuring delete operations with retry support
+///
+/// This operation is similar to SQL's DELETE statement. It allows you to remove
+/// rows from a dataset based on a filter predicate with automatic retry support
+/// for handling concurrent write conflicts.
+///
+/// Use the [DeleteBuilder] to construct a delete operation. For example:
+///
+/// ```
+/// # use lance::{Dataset, Result};
+/// # use lance::dataset::DeleteBuilder;
+/// # use std::sync::Arc;
+/// # async fn example(dataset: Arc<Dataset>) -> Result<()> {
+/// let new_dataset = DeleteBuilder::new(dataset, "age > 65")
+///     .conflict_retries(5)
+///     .execute()
+///     .await?;
+/// # Ok(())
+/// # }
+/// ```
+///
+#[derive(Debug, Clone)]
+pub struct DeleteBuilder {
+    dataset: Arc<Dataset>,
+    predicate: String,
+    conflict_retries: u32,
+    retry_timeout: Duration,
+}
 
-            apply_deletions(ds, &removed_row_addrs).await?
+impl DeleteBuilder {
+    /// Create a new DeleteBuilder
+    pub fn new(dataset: Arc<Dataset>, predicate: impl Into<String>) -> Self {
+        Self {
+            dataset,
+            predicate: predicate.into(),
+            conflict_retries: 10,
+            retry_timeout: Duration::from_secs(30),
         }
-    } else {
-        // No filter was applied - this shouldn't happen but treat as delete nothing
-        (Vec::new(), Vec::new())
-    };
-
-    let transaction = Transaction::new(
-        ds.manifest.version,
-        Operation::Delete {
+    }
+
+    /// Set the number of retries for conflict resolution
+    pub fn conflict_retries(mut self, retries: u32) -> Self {
+        self.conflict_retries = retries;
+        self
+    }
+
+    /// Set the timeout for retry operations
+    pub fn retry_timeout(mut self, timeout: Duration) -> Self {
+        self.retry_timeout = timeout;
+        self
+    }
+
+    /// Execute the delete operation
+    pub async fn execute(self) -> Result<Arc<Dataset>> {
+        let job = DeleteJob {
+            dataset: self.dataset.clone(),
+            predicate: self.predicate,
+        };
+
+        let config = RetryConfig {
+            max_retries: self.conflict_retries,
+            retry_timeout: self.retry_timeout,
+        };
+
+        execute_with_retry(job, self.dataset, config).await
+    }
+}
+
+/// Job that executes the delete operation
+#[derive(Debug, Clone)]
+struct DeleteJob {
+    dataset: Arc<Dataset>,
+    predicate: String,
+}
+
+/// Data returned by delete operation
+struct DeleteData {
+    updated_fragments: Vec<Fragment>,
+    deleted_fragment_ids: Vec<u64>,
+}
+
+impl RetryExecutor for DeleteJob {
+    type Data = DeleteData;
+    type Result = Arc<Dataset>;
+
+    async fn execute_impl(&self) -> Result<Self::Data> {
+        // Create a single scanner for the entire dataset
+        let mut scanner = self.dataset.scan();
+        scanner.with_row_id().filter(&self.predicate)?;
```

Comment:
We should exclusively project the row id, since we don't need to scan other columns.
```suggestion
        scanner.project(&[ROW_ID]).filter(&self.predicate)?;
```